platform-sdk/docs/proposals/TSS-Library/TSS-Library.md

```diff
@@ -0,0 +1,860 @@
+* Should we parametrize the EC curve with: [`java.security.spec.EllipticCurve`](https://docs.oracle.com/javase/1.5.0/docs/api/java/security/spec/EllipticCurve.html)? Implications?
+* What is the serialization format supported by arkworks?  Raw Key Bytes are formatted with PKCS#8 for private keys and X.509 for public keys.
+  Should we define a custom format for bytes serialized with arkworks? Should we reformat? What do we do with our custom content?
+
+
+#### Swirlds Crypto Pairings API
+##### Overview
+This API will expose general arithmetic operations to work with Bilinear Pairings and EC curves that implementations must provide.
+
+##### Public API
+
+###### `Curve`
+**Description**: Represents the different types of elliptic curves. Implementations of this API should decide if they provide any, one, or many curves.
+
+**Link**: [Curve.java](pairings-api%2FCurve.java)
+
+**Note**: This enum should list all possible types so changes in the implementation don't change the API, but as of now, the number is limited.
+
+###### `BilinearPairing`
+
+**Description**: This class provides access to each group (G₁, G₂) for a specific Pairing and the FiniteField associated with the curve.
+
+**Link**:  [BilinearPairing.java](pairings-api%2FBilinearPairing.java)
+
+###### `Field`
+**Description**: This is a factory interface responsible for creating FieldElement, which are scalars belonging to the field represented by this instance.
+
+**Link**: [Field.java](pairings-api%2FField.java)
+
+###### `FieldElement`
+
+**Description**: An interface representing an element within a field, providing methods for basic arithmetic operations.
+
+**Link**: [FieldElement.java](pairings-api%2FFieldElement.java)
+
+###### `Group`
+
+**Description**: This class provides methods to obtain elements belonging to the group represented by the instance.
+
+**Link**: [Group.java](pairings-api%2FGroup.java)
+
+###### `GroupElement`
+
+**Description**: An interface representing an element within a group, providing methods for basic group operations.
+
+**Link**: [GroupElement.java](pairings-api%2FGroupElement.java)
+
+###### `PairingResult`
+
+**Description**: An interface representing the result of a pairing operation, with methods to compare it to other group elements.
+
+**Link**: [PairingResult.java](pairings-api%2FPairingResult.java)
+
+
+#### Swirlds Crypto Pairings Impl
+##### Overview
+This module will be a multilanguage module following the structure:
+```
+ swirlds-crypto-pairings-impl-X
+ |_src
+    |_main
+    | |_java
+    | | |_com...
+    | |_rust
+    |   |_...
+    |_test
+      |_java
+      | |_com...
+      |_rust
+        |_...
+```
+Support for release engineering will be needed.
+This module should include Cargo for compiling rust source to binary files,
+It should be built in a way that the rust code is compiled for every supported architecture and only after that, packaged into the jar following the structure:
```

Comment:
This was decided in a meeting on June 4th with @nathanklick and @rbarkerSL.


More in detail, the reasoning behind this decision is to combine all things with the same reason to change. 
Given that we are using JNI, it requires code in both languages (java and Rust) that will change together. If we separate that unit of change into different modules, we will produce difficult workflows for developers.

Indeed, a Windows machine cannot be used to build for other architectures unless docker is involved. The proposal also mentions that on a developer machine, the produced jar will have only the version for that machine. This is a significant change considering our built logic today.
However, the artifact built and published to Maven should support all libraries, so it is transparent that dev is only using the libraries and not modifying them.

Reply:
I still don't understand how you will build an artefact containing native compilations for n architectures and operation systems in justst, that should be one step on one machine. At least the solution should be described here in a general way. With that definition I would currently vote -1 since I have no clue how that can be done without way too much complexity.

Reply:
After discussing, we agreed that things beyond the proposal's scope should not be detailed at a deep level.
I re-clarified that other teams will provide that part, so the document is clearer about that point.